src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,196 @@ internal static partial class JsonWriterHelper
+                {
+                    Debug.Assert((ptr + idx) <= (ptr + value.Length));
+                    if (NeedsEscaping(*(ptr + idx)))
+                    {
+                        goto Return;
+                    }
+                }
 
-        Return:
-            return idx;
+                idx = -1; // all characters allowed
+
+            Return:
+                return idx;
+            }
         }
 
         public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
-            int idx;
-
-            // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
-            // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
-            if (encoder != null && !value.IsEmpty)
+            fixed (char* ptr = value)
             {
-                fixed (char* ptr = value)
+                int idx = 0;
+
+                // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
+                // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
+                if (encoder != null && !value.IsEmpty)
                 {
                     idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
+                    goto Return;
                 }
-                goto Return;
-            }
 
-            for (idx = 0; idx < value.Length; idx++)
-            {
-                if (NeedsEscaping(value[idx]))
+#if BUILDING_INBOX_LIBRARY
+                if (Sse2.IsSupported)
                 {
-                    goto Return;
+                    short* startingAddress = (short*)ptr;
+                    while (value.Length - 8 >= idx)
+                    {
+                        Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + value.Length - 8));
+
+                        // Load the next 8 characters.
+                        Vector128<short> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                        // Check if any of the 8 characters need to be escaped.
+                        Vector128<short> mask = CreateEscapingMask(sourceValue);
+
+                        int index = Sse2.MoveMask(mask.AsByte());
+                        // If index == 0, that means none of the 8 characters needed to be escaped.
+                        // TrailingZeroCount is relatively expensive, avoid it if possible.
+                        if (index != 0)
+                        {
+                            // Found at least one character that needs to be escaped, figure out the index of
+                            // the first one found that needed to be escaped within the 8 characters.
+                            idx += BitOperations.TrailingZeroCount(index) >> 1;
+                            goto Return;
+                        }
+                        idx += 8;
+                        startingAddress += 8;
```

Comment:
More a general question (as there will be no measurable difference in this method):

Is this loop construct (incrementing both `idx` and `startingAddress`) "better" than something that just increments one of both like
```c#
short* startingAddress = (short*)ptr;
short* endAddress = startingAddress + value.Length - 8;

while (startingAddress < endAddress)
{
    Vector128<short> sourceValue = Sse2.LoadVector128(startingAddress);

    // ...

    startingAddress += 8
}

idx = (int)(endAddress - startingAddress);   // wherever needed
```
or just incrementing the index and derefencing with the index
```c#
short* startingAddress = (short*)ptr;

while (value.Length - 8 >= idx)
{
    Vector128<short> sourceValue = Sse2.LoadVector128(startingAddress + idx);

    // ...

    idx += 8;
}
```
Here in corefx all patterns are used, so I'm interested which one is prefered -- besides the fact that the difference for real-world-usage-loops will be within noise, the cpu has multiple execution ports, etc.

Reply:
Like you said, I don't think it matters all that much. I just wrote it that way because it was easier for me to reason about the bounds and I didn't have to update `idx` wherever it was being used, so the difference probably comes down to preference. But the alternatives you suggested would work fine as well, and on a different day I may have written it that way.